### ISSUE TITLE
directives not running on render of html

### ISSUE THREAD

**Issue by jesusfreak3** [fraywing/textAngular#1204]
Hey Guys, 

So I know you do not directly support directives, and to pass custom attributes on tags you need to edit the TextAngular Sanitize code to support what ever custom tags/attributes you are passing through like you did for ta-insert-video.  I have done this part to pass my attribute image-colorbox on the a tag.  And this works as I see the attribute on the anchor on the rendered html page.  However, my directive is never initialized on load of the page.

From all the reading, I was pretty sure inline directives should work on the initial load, but no matter what I try my directive is never initialized or called when the textangular is rendered.  This directive works everywhere else I use it so I know the directive is fine.

Any help would be amazing thanks!
Jon

**Comment by pbassut** [comment]
Hey,
So, as you said we don't support directives. I don't think I quite understood your problem so a working plunkr would be very much welcome. Then we can work together to solve your problem.

**Comment by stevokk** [comment]
I might not be following correctly but I believe you are trying to insert html directly into your page via a variable binding and expecting angular to parse this as it does with it's templates?

See this answer on how Angular handles the parsing of HTML: http://stackoverflow.com/a/18157958

**Comment by jesusfreak3** [comment]
Sorry for the delay, been super busy.  

So this is more on the rendering side.  When ta-bind renders the html, does it trigger another digest loop so that angular can pick up any directives within the html?  I am not talking about directives within the actual editor, as I know that would be crazy trying to monitor the changes and trigger the digest cycles, but this is when you load the rendered html using ta-bind.  

Ie , this code sample below is created in the editor and saved to the db.  Then when i load the view that uses the ta-bind , it pulls the html content from db (and does the custom renders) and then renders the html. (I updated the sanitizer so that image-colorbox is allowed, so you can see it within the rendered html)

<a href='sdfsdfsdf' image-colorbox><img src='fsdfsdfs'/></a>

this directive works anywhere else in the application, except when being rendered within ta-bind.

Example, if you go here
http://area.autodesk.com/customer-stories/mirror039s-edge-catalyst-shattering-the-norm

Click on any image with in there, the image-colorbox directive kicks in and opens up the colorbox.

Is there anyway with textAngular to tell the digest cycle to run after it renders with ta-bind?

Not sure if its clear.. long day.

Jon

**Comment by jesusfreak3** [comment]
Ok, so I see why it doesnt work.  In your directive for ta-bind, when you set the element html (_setInnerHTML), you are not running compile after.  As soon as you add that it works, and I can add directives either in the editor, or in the custom render and it works.

```
var _setInnerHTML = function(newval){console.log('setInner2');
        element[0].innerHTML = newval;
        $compile(element.contents())(scope);
};
```

Having this will really make it more powerful and allow us to add custom renders that simply inject directives based on attributes etc.  

I can create a fork for this if you want?

**Comment by stevokk** [comment]
It was my understanding that ta-bind is used for the html preview panel, so if we render directives within that and switch back to the presentation view will it not parse the HTML generated by the directives and break the original input?

If you want to create a fork demoing the functionality then the maintainers of textAngular can review and decide.

If they were not to merge the change, you could create an external "plugin" that used $watch on the ta-bind variable and compile outside of textAngular.

**Comment by jesusfreak3** [comment]
True, However, there are custom renders like for youtube videos (switches images to iframed videos), and this is where I am trying to work with.  So then you can create custom renders that inject the directive markup, and then when you switch to the presentation mode with ta bind the custom renders kick in and then it gets compiled.  I will need to play around with it a bit more, it works for me right now as the system i built doesnt switch between the edit and presentation mode, as there is an admin page for editing, and then the front end page that uses ta-bind to show the rendered output, and preview just brings them to that page.

Jon

**Comment by JamesRClarke** [comment]
@jesusfreak3 
I am trying to achieve something you said that you got working for you with this piece of code
```
 var _setInnerHTML = function(newval) {
        element[0].innerHTML = newval;
        $compile(element.contents())(scope);
};
```
inside the textAngular.js file, however I get the error `$compile is not a function`.

I tried injecting $compile into the taBind directive and can't fix this error.

Any help with this would be really appreciated.

**Comment by jesusfreak3** [comment]
@JamesRClarke   
Make sure you inject it into the textAngular directive, otherwise it wont know what that is.

```
textAngular.directive("textAngular", [
	'$compile', '$timeout', 'taOptions', 'taSelection', 'taExecCommand',
	'textAngularManager', '$document', '$animate', '$log', '$q', '$parse',
	function($compile, $timeout, taOptions, taSelection, taExecCommand,
		textAngularManager, $document, $animate, $log, $q, $parse){
		return {
```

**Comment by JamesRClarke** [comment]
@jesusfreak3 
I have tried that already and am having no luck, I am trying to pass through a directive that outputs a template with an inline directive, and I cannot get textAngular to compile it so I thought your soultion may help, but I cant get it working.

Thank you very much for the quick reply though!

**Comment by jesusfreak3** [comment]
@JamesRClarke np.  Did that at least ride you have the $compile function throwing an error?

You need to also make sure you update the textAngular-santize script to allow your custom tags to pass through to be compiled.

```
function validCustomTag(tag, attrs, lkey, value){
	// catch the div placeholder for the iframe replacement
    if (tag === 'img' && attrs['ta-insert-video']){
        if(lkey === 'ta-insert-video' || lkey === 'allowfullscreen' || lkey === 'frameborder' || (lkey === 'contenteditable' && value === 'false')) return true;
    }
    
    if (tag === 'img' && attrs['image-colorbox'] !== undefined){
        if(lkey === 'image-colorbox') return true;
    } 
    
    if (tag === 'a' && attrs['image-colorbox'] !== undefined){
        if(lkey === 'image-colorbox') return true;
    }
    return false;
}
```

But I also do not know exactly what you are doing.

Thanks

**Comment by JamesRClarke** [comment]
Hi @jesusfreak3, no luck again, although this time I get the error `$compile is not defined`  

```
var _setInnerHTML = function(newval) {
        element[0].innerHTML = newval;
        $compile(element.contents())(scope);
};
```

I have the tag that I'm trying to parse in the validCustomTag function already.

Thanks
